R/trackhub.R

```diff
@@ -0,0 +1,633 @@
+
+setMethod("length", "TrackHub", function(x) length(names(x)))
+
+setMethod("show", "TrackHub", function(object) {
+    cat(class(object), "repository\nuri:", uri(object), "\n")
+    cat(S4Vectors:::labeledLine("genomes", genome(object)))
+})
+
+TrackHub <- function(uri, create = FALSE) {
+    if (!isTRUEorFALSE(create))
+        stop("'create' must be TRUE or FALSE")
+    if (create) {
+        if (uriExists(uri)) {
+            message("NOTE: '", uri, "' already exists")
+            create <- FALSE
+        } ## must create this before calling normURI (requires existence)
+        else createResource(uri, dir = TRUE)
+    }
+    th <- new("TrackHub", uri = normURI(uri))
+    if (create)
+        createResource(hubFile(th))
+    th
+}
+
+setAs("character", "TrackHub", function(from) TrackHub(from))
+
+### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+### TrackHubGenome class
+###
+
+setGeneric("addGenomesField", function(x, key, value) standardGeneric("addGenomesField"))
+setGeneric("cols", function(x) standardGeneric("cols"))
+
+setClass("TrackHubGenome",
+         representation(trackhub = "TrackHub",
+                        genome = "character"),
+         contains = "TrackDb")
+
+trackhub <- function(x) x@trackhub
+
+getTrackDbContent <- function(x) {
+    Tracks <- TrackContainer()
+    contentDf <- readAndSanitize(x)
+    contentDf$V1 <- gsub("^(\\t)+", "", contentDf$V1)
+    tracksIndex <- grep("track", contentDf$V1)
+    totalTracks <- length(tracksIndex)
+    tracksIndex[[length(tracksIndex) + 1]] <- length(contentDf$V1) + 1 # to read last track from file
+    trackNo <- 1L
+    position <- 1L
+    # to speed up, reading track by track
+    while(trackNo <= totalTracks) {
+        startPosition <- tracksIndex[trackNo]
+        endPosition <- tracksIndex[trackNo + 1] - 1
+        # dynamically creating track object
+        track <- paste0(contentDf$V1[startPosition:endPosition],
+                        "= '", contentDf$V2[startPosition:endPosition], "'", collapse=",")
+        track <- paste0("Track(", track, ")")
+        track <- eval(parse(text = track))
+        Tracks[[position]] <- track
+        position <- position + 1
+        trackNo <- trackNo + 1
+    }
+    Tracks
+}
+
+createTrackHubGenome <- function(x, genomeRecord) {
+    hubContent <- getHubContent(hubFile(trackhub(x)))
+    genomesFilePath <- combineURI(trackhub(x), hubContent["genomesFile"])
+    if (uriExists(genomesFilePath) && genome(x) %in% genome(trackhub(x))) {
+        message("NOTE: Genome '", genome(x), "' already exists")
+        return ()
+    }
+    createResource(uri(x), dir = TRUE)
+    createResource(genomesFilePath)
+    setGenomeContentList(genomeRecord, genomesFilePath)
+}
+
+setMethod("genome", "TrackHubGenome", function(x) x@genome)
+
+setMethod("uri", "TrackHubGenome", function(x)
+          paste(trimSlash(uri(trackhub(x))), genome(x), sep = "/"))
+
+setMethod("cols", "TrackHubGenome", function(x) {
```

Comment:
It's not clear what this function does from the name. Perhaps there's a more descriptive name?